### ISSUE TITLE
Use idiomatic types for `filename` parameters

### ISSUE THREAD

**Issue by gkoz** [gtk-rs/gir#195]
Upgrade the "upcasts" feature to support taking `Fundamental::Filename` as
```rust
fn select_filename<T: AsRef<Path>>(&self, filename: T) -> bool
```

**Comment by EPashkin** [comment]
We can change https://github.com/gtk-rs/gir/blob/master/src/analysis/rust_type.rs#L52
to `std::path::Path` and `std::path::PathBuf`,
then change result of https://github.com/gtk-rs/gir/blob/master/src/analysis/needed_upcast.rs#L3 to enum (for AsRef).
But I have no idea what do with long double-dotted names and imports.

**Comment by gkoz** [comment]
`AsRef<std::path::Path>` might be okay at first.

One more difficulty is the need to call to `as_ref`, e.g.

```rust
filename.as_ref().to_glib_none().0
```

**Comment by gkoz** [comment]
A similar issue is `Gio.File` ([used here](https://developer.gnome.org/gtk3/stable/GtkAppChooserDialog.html#gtk-app-chooser-dialog-new) for instance) which is similar enough to `PathBuf` that it could be nice to have

```rust
pub fn new<T: IsA<Window>, U: Into<gio::File>>(parent: Option<&T>, flags: DialogFlags, file: U) -> AppChooserDialog
```

This conflicts with taking objects by reference though...

**Comment by gkoz** [comment]
Implemented in #196
